baybe/campaign.py

```diff
@@ -261,6 +269,64 @@ def recommend(
 
         return rec
 
+    def posterior(self, candidates: pd.DataFrame) -> Posterior:
+        """Get the posterior predictive distribution for the given candidates.
+
+        The predictive distribution is based on the surrogate model of the last used
+        recommender.
+
+        Args:
+            candidates: The candidate points in experimental recommendations.
+                For details, see :meth:`baybe.surrogates.base.Surrogate.posterior`.
+
+        Raises:
+            IncompatibilityError: If the underlying surrogate model exposes no
+                method for computing the posterior distribution.
+
+        Returns:
+            Posterior: The corresponding posterior object.
+            For details, see :meth:`baybe.surrogates.base.Surrogate.posterior`.
+        """
+        surrogate = self.get_surrogate()
+        if not hasattr(surrogate, method_name := "posterior"):
```

Comment:
@Scienfitz, @AVHopp: Right now, we don't enforce a `posterior` method via the `SurrogateProtocol`, hence the explicit runtime check. Since the work on the `SurrogateProtocol` is not yet finished (i.e. we still need the user guide examples for custom surrogates), I decided not to impose it for now and rather added a TODO note in the protocol class. I think once we've agreed on the final layout, these things can be easily resolved

Reply:
Yes, agree. If we stick with the current purpose of the protocol, then `posterior` should not be included. In fact, if you think about it, there is no reason not to have several protocols in the future if we need them, e.g. one defining the requirements from the optimization point of view, the other from the diagnostics point of view, etc.

My question was actually more in the direction of `what do we want the optimization protocol to look like`, because even there we have several choices. Ultimately, this defines the essence of the interface we impose on the user, so it's probably worth to think for it for a little while. But I guess we'll see more of the pros/cons when we actually write the new examples for custom surrogates. Then, we'll notice if the current interface is actually worth anything and makes life simpler.

Just to give you ONE arbitrary example how we could change the status quo if we wanted (not saying that I want):
Instead of imposing
```python 
def fit(
    self,
    searchspace: SearchSpace,
    objective: Objective,
    measurements: pd.DataFrame,
) -> None: ...

def to_botorch(self) -> Model: ...
``` 

we could just as well encapsulate this a single requirement

```python 
def to_botorch(
    self,
    searchspace: SearchSpace,
    objective: Objective,
    measurements: pd.DataFrame,
) -> Model: ...
```

And there are probably even other possible directions. But I think we agree up to the current point and can revise when updating the user guide. Makes all sense to you?

Reply:
agree